### PR TITLE
Registration

### PR DESCRIPTION
**Pull request by Allcharles**
# Registration

Enable the use of registration on the website

## Changes

- Added recatpcha to `FormComponent`
- Updated `SecurityService` to handle registering users
- Updated `RegisterComponent` to use new behaviour

## Problems

- [ ] Need to complete unit tests (awaiting architecture review first)

## Issues

Relevant to #610
Relevant to #356

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: src/app/components/security/pages/register/register.component.ts [hunk 1]
```diff
@@ -1,53 +1,82 @@
...
+
     if (this.api.isLoggedIn()) {
       // Disable submit button
       this.loading = true;
       this.notifications.error("You are already logged in.");
+      return;
     }
+
+    this.api
+      .signUpSeed()
```
Comment (atruskie) [inline src/app/components/security/pages/register/register.component.ts:72]: we'll probably want to standardize this call to `new()` once we remove the hacks (we'll still need recapcha seeds even with functional REST APIs). Not important for now though.

File: src/app/components/security/pages/register/register.component.ts [hunk 2]
```diff
@@ -1,53 +1,82 @@
...
   `,
 })
-class RegisterComponent extends PageComponent implements OnInit {
-  public model = {};
+class RegisterComponent
+  extends FormTemplate<RegisterDetails>
+  implements OnInit {
   public fields = fields;
   public loading: boolean;
+  public seed: string | boolean = true;
```
Comment (atruskie) [inline src/app/components/security/pages/register/register.component.ts]: it is a little unclear what this seed is - I'm assuming recaptcha?

File: src/app/components/shared/form/form.component.ts [hunk 3]
```diff
@@ -31,31 +35,68 @@ export class FormComponent extends withUnsubscribe() implements OnInit {
   @Input() public submitLoading: boolean;
```
Comment (atruskie) [inline src/app/components/shared/form/form.component.ts:36]: why is this an input?

File: src/app/components/shared/form/form.component.ts [hunk 4]
```diff
@@ -31,31 +35,68 @@ export class FormComponent extends withUnsubscribe() implements OnInit {
...
   @Input() public subTitle?: string;
   @Input() public title?: string;
+  /**
+   * Recaptcha seed, using the following rules:
+   * string: Seed value
+   * true: Will set seed
+   * false: Will not set seed
+   */
+  @Input() public seed?: string | boolean = false;
+  private loadingSeed: boolean;
```
Comment (atruskie) [inline src/app/components/shared/form/form.component.ts]: these value definitions feel a bit icky. Maybe

undefined - will not use recaptcha
"LOADING" - waitng for recpatcha seed
string:  - seed available

Actually, straight from the TS  handbook: https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions

```
type RecaptchaLoadingState = {
  state: "loading";
};
type RecaptchaLoadedState = {
  state: "loaded";
  seed: string;
};
type RecaptchaState = RecaptchaLoadingState | RecaptchaLoadedState ;
```

Then type declartion is `RecaptchaState`

Comment (atruskie) [reply src/app/components/shared/form/form.component.ts]: suggested name: `recaptchaSeed`

**Review by atruskie — APPROVED** [review]
Looks great - I'd suggest making the `seed` name a bit clearer though

File: src/app/components/shared/form/form.component.ts (hunk 3 above)
Comment (Allcharles) [reply src/app/components/shared/form/form.component.ts:36]: Allows the parent component to disable the submit button while it is loading the response from the submission

**Pull request merged**